### DISCUSSION TITLE
I can't update status object within an object ticket

### DISCUSSION THREAD

**Discussion by SafaeZer** [symfony/symfony#45458] [Q&A] [answered]
I need to update the status field in ticket table but it always return the old value and I can't deal with it
This is my controller
`/**
     * @Route("/support-progress/{ticket}", name="support_progress", methods={"GET","POST"}, requirements={"ticket"="\d+"})
     */
    public function progress(Request $request, Ticket $ticket): Response
    {
        $this->denyAccessUnlessGranted('ROLE_SUPPORT');

        $status = $ticket->getStatus();

        $form = $this->createForm(ProgressType::class, $status);
        $form->handleRequest($request);

        if ($form->isSubmitted() && $form->isValid()) {
            $ticket->setStatus($status);
            $this->em->flush();

            return $this->redirectToRoute('support');
        }

        return $this->render('support/progress.html.twig', [
            'form' => $form->createView(),
        ]);
    }`

This is my Form 
`class ProgressType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {

        $builder

            ->add('status', EntityType::class, [
                'class' => Status::class,
                'mapped' => false,
                'placeholder' => '-- select an option --',
                'choice_label'  => fn (Status $status) => $status->getName(),

            ]);
    }

    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'data_class' => Status::class,
        ]);
    }
}`

**Comment by michaoj** [comment]
This happens because you retrieve the old value with this line

`$status = $ticket->getStatus();`

And then set it again

`$ticket->setStatus($status);`

If you remove this line, it should work better.

If you need more explanations refer to [https://symfony.com/doc/current/forms.html#processing-forms](https://symfony.com/doc/current/forms.html#processing-forms)

**Reply by SafaeZer** [reply]
even when I removed the line, nothing changed it stays the same value which is "New"

**Comment by michaoj** [comment]
Try to get the value from `$form->getData()` and then set it

**Reply by SafaeZer** [reply]
I did that as well but it gets only "New" value

**Comment by chyshyya** [comment, marked as answer]
In this case:
```
-$ticket->setStatus($status);
+$this->em->persist($status);
$this->em->flush();
```
Because you didn't change $ticket at all, but do changes inside Status entity

**Reply by SafaeZer** [reply]
I did what you've said but it didn't change

**Reply by chyshyya** [reply]
Just see that status form has mapped false. Thats why object not updated. My first sugession must be applied too.
From docs:
[mapped](https://symfony.com/doc/current/reference/forms/types/form.html#mapped)
type: boolean default: true
If you wish the field to be **_ignored when reading or writing to the object_**, you can set the mapped option to false.

**Reply by chyshyya** [reply]
remove mapped or set it to true

**Reply by SafaeZer** [reply]
if I set it to true or remove it I get this Error:  Can't get a way to read the property "status" in class "Proxies\__CG__\App\Entity\Status".

**Reply by chyshyya** [reply]
I think some logic is a problem.

Write your symfony and php versions.
Drag & drop some files here:
1. Ticket entity and Repository
2. Status entity and Repository
3. support/progress.html.twig template file
I'll test them in my dev environment and try to find out what's wrong

**Reply by chyshyya** [reply]
As for me I think that Ticket has field status and you wish to change it.
So form default data_class must be Ticket

```
    $this->denyAccessUnlessGranted('ROLE_SUPPORT');

    //-$status = $ticket->getStatus();

    //-$form = $this->createForm(ProgressType::class, $status);
    $form = $this->createForm(ProgressType::class, $ticket);
    $form->handleRequest($request);

    if ($form->isSubmitted() && $form->isValid()) {
        //-$ticket->setStatus($status);
        $this->em->flush();

        return $this->redirectToRoute('support');
    }

    return $this->render('support/progress.html.twig', [
        'form' => $form->createView(),
    ]);
```
And form:
```
class ProgressType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder
            ->add('status', EntityType::class, [
                'class' => Status::class,
                //-'mapped' => false,
                'placeholder' => '-- select an option --',
                'choice_label'  => fn (Status $status) => $status->getName(),
            ]);
    }
    
    public function configureOptions(OptionsResolver $resolver): void
    {
        $resolver->setDefaults([
            'data_class' => Ticket::class, // we work with status field in Ticket entity
        ]);
    }
}
```

**Reply by SafaeZer** [reply]
Ticket Entity :
`namespace App\Entity;

use App\Repository\TicketRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Gedmo\Mapping\Annotation as Gedmo;

/**
 * @ORM\Entity(repositoryClass=TicketRepository::class)
 */
class Ticket
{
    /**
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     */
    private $id;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $titre;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $description;

    /**
     * @ORM\Column(type="string", nullable=true)
     */
    private $files;

    /**
     * @ORM\Column(type="datetime")
     * @Gedmo\Timestampable(on="create")
     */
    private $created_date;

    /**
     * @ORM\Column(type="datetime")
     * @Gedmo\Timestampable(on="update")
     */
    private $updated_date;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $priority;

    /**
     * @ORM\ManyToOne(targetEntity=TicketType::class, inversedBy="tickets")
     * @ORM\JoinColumn(nullable=false)
     */
    private $ticket_type;

    /**
     * @ORM\ManyToOne(targetEntity=Category::class, inversedBy="tickets")
     * @ORM\JoinColumn(nullable=false)
     */
    private $category;


    /**
     * @ORM\ManyToOne(targetEntity=Status::class, inversedBy="tickets")
     * @ORM\JoinColumn(nullable=false)
     */
    private $status;

    /**
     * @ORM\Column(type="boolean", nullable=true)
     */
    private $state;

    /**
     * @ORM\ManyToOne(targetEntity=User::class, inversedBy="tickets")
     */
    private $responsible;

    /**
     * @ORM\ManyToOne(targetEntity=User::class, inversedBy="tickets")
     */
    private $author;





    public function getId(): ?int
    {
        return $this->id;
    }

    public function getTitre(): ?string
    {
        return $this->titre;
    }

    public function setTitre(string $titre): self
    {
        $this->titre = $titre;

        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(string $description): self
    {
        $this->description = $description;

        return $this;
    }

    public function getFiles()
    {
        return $this->files;
    }

    public function setFiles($files)
    {
        $this->files = $files;

        return $this;
    }

    public function getCreatedDate(): ?\DateTimeInterface
    {
        return $this->created_date;
    }



    public function getUpdatedDate(): ?\DateTimeInterface
    {
        return $this->updated_date;
    }



    public function getPriority(): ?string
    {
        return $this->priority;
    }

    public function setPriority(string $priority): self
    {
        $this->priority = $priority;

        return $this;
    }

    public function getTicketType(): ?TicketType
    {
        return $this->ticket_type;
    }

    public function setTicketType(?TicketType $ticket_type): self
    {
        $this->ticket_type = $ticket_type;

        return $this;
    }

    public function getCategory(): ?Category
    {
        return $this->category;
    }

    public function setCategory(?Category $category): self
    {
        $this->category = $category;

        return $this;
    }

    public function getStatus(): ?Status
    {
        return $this->status;
    }

    public function setStatus(?Status $status): self
    {
        $this->status = $status;

        return $this;
    }

    public function getState(): ?bool
    {
        return $this->state;
    }

    public function setState(?bool $state): self
    {
        $this->state = $state;

        return $this;
    }
    public function __toString()
    {
        return (string) $this;
    }
public function getResponsible(): ?User
    {
        return $this->responsible;
    }

    public function setResponsible(?User $responsible): self
    {
        $this->responsible = $responsible;

        return $this;
    }

    public function getAuthor(): ?User
    {
        return $this->author;
    }

    public function setAuthor(?User $author): self
    {
        $this->author = $author;

        return $this;
    }}`
Ticket Repository :
`namespace App\Repository;

use App\Entity\Ticket;
use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
use Doctrine\Persistence\ManagerRegistry;

/**
 * @method Ticket|null find($id, $lockMode = null, $lockVersion = null)
 * @method Ticket|null findOneBy(array $criteria, array $orderBy = null)
 * @method Ticket[]    findAll()
 * @method Ticket[]    findBy(array $criteria, array $orderBy = null, $limit = null, $offset = null)
 */
class TicketRepository extends ServiceEntityRepository
{
    public function __construct(ManagerRegistry $registry)
    {
        parent::__construct($registry, Ticket::class);
    }

    // /**
    //  * @return Ticket[] Returns an array of Ticket objects
    //  */
    /*
    public function findByExampleField($value)
    {
        return $this->createQueryBuilder('t')
            ->andWhere('t.exampleField = :val')
            ->setParameter('val', $value)
            ->orderBy('t.id', 'ASC')
            ->setMaxResults(10)
            ->getQuery()
            ->getResult()
        ;
    }
    */

    /*
    public function findOneBySomeField($value): ?Ticket
    {
        return $this->createQueryBuilder('t')
            ->andWhere('t.exampleField = :val')
            ->setParameter('val', $value)
            ->getQuery()
            ->getOneOrNullResult()
        ;
    }
    */
}`
Status Entity :
`<?php

namespace App\Entity;

use App\Repository\StatusRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity(repositoryClass=StatusRepository::class)
 */
class Status
{
    /**
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     */
    private $id;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $name;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $color;

    /**
     * @ORM\OneToMany(targetEntity=Ticket::class, mappedBy="status")
     */
    private $tickets;



    public function __construct()
    {
        $this->tickets = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(string $name): self
    {
        $this->name = $name;

        return $this;
    }

    public function getColor(): ?string
    {
        return $this->color;
    }

    public function setColor(string $color): self
    {
        $this->color = $color;

        return $this;
    }

    /**
     * @return Collection|Ticket[]
     */
    public function getTickets(): Collection
    {
        return $this->tickets;
    }

    public function addTicket(Ticket $ticket): self
    {
        if (!$this->tickets->contains($ticket)) {
            $this->tickets[] = $ticket;
            $ticket->setStatus($this);
        }

        return $this;
    }

    public function removeTicket(Ticket $ticket): self
    {
        if ($this->tickets->removeElement($ticket)) {
            // set the owning side to null (unless already changed)
            if ($ticket->getStatus() === $this) {
                $ticket->setStatus(null);
            }
        }

        return $this;
    }
    public function __toString()
    {
        return (string) $this->name;
    }
}`
Status Repository : 
`namespace App\Repository;

use App\Entity\Status;
use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
use Doctrine\Persistence\ManagerRegistry;

/**
 * @method Status|null find($id, $lockMode = null, $lockVersion = null)
 * @method Status|null findOneBy(array $criteria, array $orderBy = null)
 * @method Status[]    findAll()
 * @method Status[]    findBy(array $criteria, array $orderBy = null, $limit = null, $offset = null)
 */
class StatusRepository extends ServiceEntityRepository
{
    public function __construct(ManagerRegistry $registry)
    {
        parent::__construct($registry, Status::class);
    }

    // /**
    //  * @return Status[] Returns an array of Status objects
    //  */
    /*
    public function findByExampleField($value)
    {
        return $this->createQueryBuilder('s')
            ->andWhere('s.exampleField = :val')
            ->setParameter('val', $value)
            ->orderBy('s.id', 'ASC')
            ->setMaxResults(10)
            ->getQuery()
            ->getResult()
        ;
    }
    */

    /*
    public function findOneBySomeField($value): ?Status
    {
        return $this->createQueryBuilder('s')
            ->andWhere('s.exampleField = :val')
            ->setParameter('val', $value)
            ->getQuery()
            ->getOneOrNullResult()
        ;
    }
    */
}`
support/progress.html.twig Template:
`{% extends '/support/support.html.twig' %}

{% block body %}
	{{ form_start(form) }}
	<div class="container mt-5">
		<div class="col-6 mx-auto">
			{{ form_widget(form) }}
			<button class="btn btn-success">{{ button_label|default('Mark') }}</button>
		</div>
	</div>
	{{ form_end(form) }}


{% endblock %}`

**Reply by chyshyya** [reply]
Try my previous comment. I think it will do the job for you.

**Reply by chyshyya** [reply]
The main thing is not to miss to change Status to Ticket
```
        $resolver->setDefaults([
            'data_class' => Ticket::class, // we work with status field in Ticket entity
        ]);
```

**Reply by SafaeZer** [reply]
It's working now Thank you so much @chyshyya I really appreciate yr help and time you saved my day ^_^
